### ISSUE TITLE
Generate compile_commands.json with dry run

### ISSUE THREAD

**Issue by neikeq** [godotengine/godot#39883]
**TL;DR:** We need a way to generate a compilation database without having to build Godot.

Right now our SCons script generates a `compile_commands.json` when building the project. I've been using this for some time for better support in IDEs. I was using it for QtCreator's ClangCodeModel and I'm now using it with CLion.

However, there's a problem: in order for this file to be generated, we must wait for a full build of Godot to be completed. This means I have to do a complete build of Godot before I can start editing engine code. It worsens by the fact I'm constantly jumping from different versions of the Godot source tree, most of the time because I upgrade to latest master or because I switch between the master and 3.2 branches.

I tried running SCons with `--dry-run`. This options prints all the build commands, but doesn't run them. However, when running SCons with `--dry-run` the `compile_commands.json` wasn't generated.

**Comment by neikeq** [comment]
cc @RevoluPowered Any idea if this could be made to work somehow?

**Comment by RevoluPowered** [comment]
It might be possible but I am unsure how we could go about making this work, I will look into it though and get back to you sorry for the slow reply been FBX'ing!

**Comment by SlawekNowy** [comment]
It does not help that Mono version requires two compilations, should we need to build from scratch.

One solution that would be to imply dry ran full compile to compiledb target.

**Comment by kidinashell** [comment]
I know, this is more of a general question, but I haven't found anything on the matter elsewhere (docs or community q&a).
I built the whole project from the 3.2 branch but there was no compilation database generated. Is there only a specific branch, that generates compile_commands.json? Thank you in advance.

**Comment by dmoody256** [comment]
> TL;DR: We need a way to generate a compilation database without having to build Godot.

All files are valid targets to scons, to build the compile commands by itself wihtout building anything else in Godot, just supply the file as a target:
```
scons <options...> compile_commands.json
```

**Comment by neikeq** [comment]
> All output files (and directories) are valid targets to scons, to build the compile commands by itself without building anything else in Godot, just supply the file as a target:
> 
> ```
> scons <options...> compile_commands.json
> ```

That definitely solves the problem. I had no idea you could do that, thanks!

**Comment by akien-mga** [comment]
Might be worth including in https://docs.godotengine.org/en/stable/development/compiling/introduction_to_the_buildsystem.html as this might be useful to others.

**Comment by rcorre** [comment]
> > TL;DR: We need a way to generate a compilation database without having to build Godot.
> 
> All output files (and directories) are valid targets to scons, to build the compile commands by itself without building anything else in Godot, just supply the file as a target:
> 
> ```
> scons <options...> compile_commands.json
> ```

For anyone finding this later, note that it has to be `scons compiledb=yes compile_commands.json`. I opened https://github.com/godotengine/godot-docs/pull/6888 to document this.
